### PR TITLE
Subproject: bug fixes, speedups, reduced verbosity

### PR DESCRIPTION
**Pull request by tribal-tec**
@rdumusc Have fun...

### CODE REVIEW

**Comment by tribal-tec** [comment]
@hernando Please have a look at ChoosePython.cmake and its usage in Common.cmake

File: ChoosePython.cmake [hunk 1]
```diff
@@ -8,11 +8,12 @@
 # This script provides an option to choose which Python version should be
 # requested to FindPythonInterp and FindPythonLibs. The user can choose
 # between version 2, 3 or autodetection (the default). If
-# CHOOSE_PYTHON_REQUIRE_BOOST is set, autodetection will choose Python 3 only
-# if Boost.Python is found to have support for Python 3.
+# CHOOSE_PYTHON_IGNORE_BOOST is set, autodetection will choose the Python
+# version without checking if Boost.Python is available and which version it
+# does support.
```
Comment (rdumusc) [inline ChoosePython.cmake]: it supports

File: CommonLibrary.cmake [hunk 2]
```diff
@@ -11,7 +11,8 @@
 # * NAME_LIBRARY_TYPE or COMMON_LIBRARY_TYPE for SHARED or STATIC library, with
 #   COMMON_LIBRARY_TYPE being an option stored in the CMakeCache.
 # * NAME_OMIT_PROJECT_HEADER when set, no project header (name.h) is generated.
-# * PROJECT_INCLUDE_NAME for the include directory and project include header
+# * NAME_INCLUDE_NAME for the include directory and project include header
```
Comment (rdumusc) [inline CommonLibrary.cmake]: and library header (no longer for the whole project)

File: CommonLibrary.cmake [hunk 3]
```diff
@@ -36,10 +37,10 @@ set(COMMON_LIBRARY_TYPE SHARED CACHE STRING
 set_property(CACHE COMMON_LIBRARY_TYPE PROPERTY STRINGS SHARED STATIC)
 
 macro(GENERATE_PROJECT_HEADER NAME)
-  get_filename_component(BASE_NAME ${PROJECT_INCLUDE_NAME} NAME)
+  get_filename_component(BASE_NAME ${INCLUDE_NAME} NAME)
 
   set(PROJECT_GENERATED_HEADER
```
Comment (rdumusc) [inline CommonLibrary.cmake]: rename PROJECT_GENERATED_HEADER -> GENERATED_HEADER

File: CommonLibrary.cmake [hunk 4]
```diff
@@ -54,12 +55,12 @@ macro(GENERATE_PROJECT_HEADER NAME)
...
-        "#include <${PROJECT_INCLUDE_NAME}/${PUBLIC_HEADER}>\n")
+        "#include <${INCLUDE_NAME}/${PUBLIC_HEADER}>\n")
     endif()
   endforeach()
   file(APPEND ${PROJECT_GENERATED_HEADER} "#endif\n")
   set(PROJECT_INCLUDE_HEADER
-    ${PROJECT_BINARY_DIR}/include/${PROJECT_INCLUDE_NAME}/${BASE_NAME}.h)
+    ${OUTPUT_INCLUDE_DIR}/${INCLUDE_NAME}/${BASE_NAME}.h)
 
   configure_file(
```
Comment (rdumusc) [inline CommonLibrary.cmake]: Why is this done in two steps? I see no differences between the .in and the .h?

Comment (tribal-tec) [inline CommonLibrary.cmake]: Good point. I don't know why. I accidentally inherited this bad code :) I'll fix it and test.

File: CommonLibrary.cmake (hunk 3 above)
Comment (tribal-tec) [inline CommonLibrary.cmake]: I'll rename everything from here which starts with PROJECT_

File: CommonPackage.cmake [hunk 5]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
...
+set(FIND_PACKAGES_DEFINES ${SYSTEM})
+
+macro(common_package Name)
   string(TOUPPER ${Name} COMMON_PACKAGE_NAME)
   set(COMMON_PACKAGE_ARGS ${ARGN}) # ARGN is not a list. make one.
-  set(COMMON_PACKAGE_VERSION)
 
+  # SYSTEM specified?
+  if(COMMON_PACKAGE_ARGS)
+    string(REGEX MATCH "SYSTEM?" COMMON_PACKAGE_SYSTEM ${COMMON_PACKAGE_ARGS})
```
Comment (rdumusc) [inline CommonPackage.cmake]: for readability, rename COMMON_PACKAGE_SYSTEM to IS_SYSTEM_PACKAGE

File: CommonPackage.cmake [hunk 6]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
...
+  if(${Name} STREQUAL "Boost")
+    set(COMMON_PACKAGE_SYSTEM ON)
+  endif()
+
+  # COMPONENTS specified?
   if(COMMON_PACKAGE_ARGS)
-    list(GET COMMON_PACKAGE_ARGS 0 COMMON_PACKAGE_VERSION)
-    if(COMMON_PACKAGE_VERSION MATCHES "^[0-9.]+$") # is a version
-      set(COMMON_PACKAGE_VERSION ">=${COMMON_PACKAGE_VERSION}")
+    string(REGEX MATCH "COMPONENTS?" COMMON_PACKAGE_COMPONENTS ${COMMON_PACKAGE_ARGS})
```
Comment (rdumusc) [inline CommonPackage.cmake]: same here, rename COMMON_PACKAGE_COMPONENTS to PACKAGE_HAS_COMPONENTS

File: CommonPackage.cmake [hunk 7]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
...
-    set(COMMON_PACKAGE_FIND_QUIET "QUIET")
-  else()
-    list(FIND COMMON_PACKAGE_ARGS "QUIET" COMMON_PACKAGE_QUIET_POS)
-    if(COMMON_PACKAGE_QUIET_POS EQUAL -1)
-      set(COMMON_PACKAGE_FIND_QUIET)
-    else()
+  # OPT: only forward to find_package if not found or component find (tbd
+  # implement found_components check)
+  if((NOT ${Name}_FOUND AND NOT ${COMMON_PACKAGE_NAME}_FOUND) OR COMMON_PACKAGE_COMPONENTS)
+    set(COMMON_PACKAGE_VERSION)
```
Comment (rdumusc) [inline CommonPackage.cmake]: COMMON_PACKAGE_VERSION -> PACKAGE_VERSION ?

File: CommonPackage.cmake [hunk 8]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
...
+    else()
+      list(FIND COMMON_PACKAGE_ARGS "QUIET" COMMON_PACKAGE_QUIET_POS)
+      if(COMMON_PACKAGE_QUIET_POS EQUAL -1)
+        set(COMMON_PACKAGE_FIND_QUIET)
+      else()
+        set(COMMON_PACKAGE_FIND_QUIET "QUIET")
+      endif()
+    endif()
+
+    list(FIND COMMON_PACKAGE_ARGS "REQUIRED" COMMON_PACKAGE_REQUIRED_POS)
```
Comment (rdumusc) [inline CommonPackage.cmake]: COMMON_PACKAGE_REQUIRED_POS -> PACKAGE_IS_REQUIRED?

File: CommonPackage.cmake [hunk 9]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
 
 option(COMMON_PACKAGE_QUIET "Use QUIET for common_package command" ON)
 
-macro(COMMON_PACKAGE Name)
+include(System)
+set(FIND_PACKAGES_DEFINES ${SYSTEM})
+
+macro(common_package Name)
   string(TOUPPER ${Name} COMMON_PACKAGE_NAME)
```
Comment (rdumusc) [inline CommonPackage.cmake]: PACKAGE_NAME?

File: CommonPackage.cmake [hunk 10]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
...
-    find_package(${Name} ${COMMON_PACKAGE_FIND_QUIET} ${COMMON_PACKAGE_ARGS}) # try standard cmake way
-    if((NOT ${Name}_FOUND) AND (NOT ${COMMON_PACKAGE_NAME}_FOUND) AND PKG_CONFIG_EXECUTABLE)
-      pkg_check_modules(${Name} ${Name}${COMMON_PACKAGE_VERSION}
-        ${COMMON_PACKAGE_FIND_QUIET}) # try pkg_config way
+  if(EXISTS ${PROJECT_SOURCE_DIR}/CMake/FindPackagesPost.cmake)
+    include(${PROJECT_SOURCE_DIR}/CMake/FindPackagesPost.cmake)
+  endif()
+
+  string(TOUPPER ${PROJECT_NAME} UPPER_PROJECT_NAME)
+  set(DEFDEP "${UPPER_PROJECT_NAME}_USE_${COMMON_PACKAGE_NAME}")
```
Comment (rdumusc) [inline CommonPackage.cmake]: move DEFDEF down to where it is used (line 115). if possible, find a more meaningful name :-)

File: CommonPackage.cmake [hunk 11]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
...
+    set(${Name}_name ${COMMON_PACKAGE_NAME})
+    set(${Name}_FOUND TRUE)
+  elseif(${Name}_FOUND)
+    set(${Name}_name ${Name})
+    set(${COMMON_PACKAGE_NAME}_FOUND TRUE)
+  else()
+    set(${PROJECT_NAME}_FIND_PACKAGES_NOTFOUND "${${PROJECT_NAME}_FIND_PACKAGES_NOTFOUND} ${Name}")
+  endif()
+  if(${Name}_name)
+    list(APPEND FIND_PACKAGES_DEFINES ${DEFDEP})
```
Comment (rdumusc) [inline CommonPackage.cmake]: The fact that common_package accumulates values in FIND_PACKAGES_DEFINES is not documented

File: README.md [hunk 12]
```diff
@@ -23,14 +23,17 @@ the beginning of the CMake run, and make all its features available.
 The following CMake modules can be included in your project:
 
 * [Common](Common.cmake) does a common CMake setup, and also includes:
-    * [CommonCode](CommonCode.cmake) generates standard header files for
-      the project (api.h, defines.h, version.h).
     * [CommonLibrary](CommonLibrary.cmake) *common_library* function to
-      build a shared library using a standard recipe.
+      build a shared library using a standard recipe and generates header
+      files for the library (api.h, defines.h, version.h).
```
Comment (rdumusc) [inline README.md]: I think common_package_post() generates "defines.h" now, right?

File: System.cmake [hunk 13]
```diff
@@ -1,16 +1,20 @@
 # sets ${SYSTEM}
```
Comment (rdumusc) [inline System.cmake]: Improve documentation?
Output:
* SYSTEM, a human-readable string identifier for the current platform (Win32, Darwin or Linux)
* OSX_VERSION, the two numbers OS X version (e.g. 10.9)

File: CommonPackage.cmake [hunk 14]
```diff
@@ -14,44 +25,181 @@ set(ENV{PKG_CONFIG_PATH}
...
+    message(FATAL_ERROR "PROJECT_INCLUDE_NAME not set, old or missing Common.cmake?")
+  endif()
+  if(NOT OPTIONS_CMAKE)
+    set(OPTIONS_CMAKE ${CMAKE_CURRENT_BINARY_DIR}/options.cmake)
+  endif()
+
+  configure_file(${CMAKE_SOURCE_DIR}/CMake/common/cpp/defines.h
+    ${OUTPUT_INCLUDE_DIR}/${PROJECT_INCLUDE_NAME}/defines.h @ONLY)
+  set(DEFINES_FILE "${CMAKE_CURRENT_BINARY_DIR}/include/${PROJECT_INCLUDE_NAME}/defines${SYSTEM}.h")
+  list(APPEND COMMON_INCLUDES
```
Comment (rdumusc) [inline CommonPackage.cmake]: Is COMMON_INCLUDES still used somewhere? If yes, it must be documented in the output variables. If not, then defines.h does not get installed anymore, despite the fact that api.h includes it.

File: CommonPackage.cmake (hunk 5 above)
Comment (tribal-tec) [inline CommonPackage.cmake]: In fact, those are all private, local variables. I'll rename with underscores

File: CommonPackage.cmake [hunk 15]
```diff
@@ -1,12 +1,23 @@
...
+#  - QUIET if COMMON_PACKAGE_QUIET option is set
+#  - -isystem if SYSTEM argument is passed; for convenience, Boost is always SYSTEM
+#  - first tries find_package with all the given arguments, and then falls back
+#    to using pkg_config if available (no component support for pkg_config though)
+#  - processes ${PROJECT_SOURCE_DIR}/CMake/FindPackagesPost.cmake to tweak result
+#    of common_package_post()
+#  - sets include_directories() and link_directories() accordingly
+#
+# common_package_post()
+#  - generates defines${SYSTEM}.h and options.cmake for found packages
```
Comment (rdumusc) [inline CommonPackage.cmake]: Clarify the difference between defines.h and defines${SYSTEM}.h? I'm not sure why there are two files. I'm also not sure how options.cmake is meant to be used...

File: CommonPackage.cmake (hunk 10 above)
Comment (tribal-tec) [inline CommonPackage.cmake]: Will do

File: CommonPackage.cmake (hunk 11 above)
Comment (tribal-tec) [inline CommonPackage.cmake]: It's more a detail which is kindly resolved by calling common_package_post(), but I'll document the in/out variables

File: CommonPackage.cmake (hunk 5 above)
Comment (rdumusc) [inline CommonPackage.cmake]: yes, perfect!

**Comment by tribal-tec** [comment]
Updated. Have a last look please.

File: CommonPackage.cmake [hunk 16]
```diff
@@ -1,57 +1,228 @@
...
+# improves find_package.
+#
+# common_package()
+#  - QUIET if COMMON_PACKAGE_USE_QUIET option is set
+#  - -isystem if SYSTEM argument is passed; for convenience, Boost is always
+#    SYSTEM
+#  - first tries find_package with all the given arguments, and then falls back
+#    to using pkg_config if available (no component support for pkg_config though)
+#  - processes ${PROJECT_SOURCE_DIR}/CMake/FindPackagesPost.cmake to tweak result
+#    of common_package_post()
```
Comment (eile) [inline CommonPackage.cmake]: This is still happening/needed?

File: CommonPackage.cmake [hunk 17]
```diff
@@ -1,57 +1,228 @@
...
+#  - sets include_directories() and link_directories() accordingly
+#
+# common_package_post()
+#  - generates defines.h and options.cmake for found packages.
+#  - prints status message of found and not-found packages
+#
+# Input variables
+#  - CMAKE_INSTALL_PREFIX - install prefix, comes from Common.cmake
+#  - CMAKE_MODULE_INSTALL_PATH -
```
Comment (eile) [inline CommonPackage.cmake]: ?

File: CommonPackage.cmake [hunk 18]
```diff
@@ -1,57 +1,228 @@
...
+#  - UPPER_PROJECT_NAME - upper-case project name, comes from Common.cmake
+#
+# Output variables
+#  - COMMON_PACKAGE_DEFINES - accumulated defines of found packages for
+#    options.cmake and defines.h, written by common_package_post()
+#  - ${PROJECT_NAME}_FIND_PACKAGES_FOUND - list of found packages
+#  - ${PROJECT_NAME}_FIND_PACKAGES_NOTFOUND - list of not-found packages
```
Comment (eile) [inline CommonPackage.cmake]: Both are not a list, but a string.

File: Compiler.cmake [hunk 19]
```diff
@@ -5,17 +5,40 @@
...
+# * COMMON_C_FLAGS The common flags for C compiler
+# * COMMON_C_FLAGS_RELWITHDEBINFO The common flags for C compiler in RelWithDebInfo build
+# * COMMON_C_FLAGS_RELEASE The common flags for C compiler in Release build
+# * COMMON_CXX_FLAGS The common flags for C++ compiler
+# * COMMON_CXX_FLAGS_RELWITHDEBINFO The common flags for C++ compiler in RelWithDebInfo build
+# * COMMON_CXX_FLAGS_RELEASE The common flags for C++ compiler in Release build
 # * C_DIALECT_OPT_C89    Compiler flag to select C89 C dialect
 # * C_DIALECT_OPT_C89EXT Compiler flag to select C89 C dialect with extensions
 # * C_DIALECT_OPT_C99    Compiler flag to select C99 C dialect
 # * C_DIALECT_OPT_C99EXT Compiler flag to select C99 C dialect with extensions
```
Comment (eile) [inline Compiler.cmake]: + BIGENDIAN?

File: CommonPackage.cmake (hunk 16 above)
Comment (tribal-tec) [inline CommonPackage.cmake]: Not needed, but quite extensive for Collage and Equalizer. Later please.

File: CommonPackage.cmake (hunk 17 above)
Comment (tribal-tec) [inline CommonPackage.cmake]: Oops, forgot that.

**Comment by rdumusc** [comment]
+1

**Comment by tribal-tec** [comment]
Yeah!

**Comment by eile** [comment]
+1

**Pull request merged**